lib/network/validation/proto/proto.go

```diff
@@ -55,14 +58,71 @@ func (r CheckBandwidthRequest) Check() error {
 	return nil
 }
 
-// Checks validates the disk performance test request.
-func (r CheckDisksRequest) Check() error {
-	if len(r.Spec) == 0 {
-		return trace.BadParameter("job spec is empty")
+// CheckAndSetDefaults validates the request and sets defaults.
+func (r *CheckDisksRequest) CheckAndSetDefaults() error {
+	for _, job := range r.Jobs {
+		if err := job.Check(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	// During operations such as install or join fio binary is placed
+	// in temporary directory on the nodes so look there if the path
+	// wasn't specified explicitly.
+	if r.FioPath == "" {
+		r.FioPath = filepath.Join(os.TempDir(), constants.FioBin)
+	}
+	if _, err := utils.StatFile(r.FioPath); err != nil {
+		return trace.Wrap(err)
 	}
 	return nil
 }
 
+// Check validates the fio job spec.
+func (r FioJobSpec) Check() error {
+	// Generally fio does not mandate filename and will generate one if it
+	// is not provided, but we require it to simplify things like cleanup.
+	if r.Filename == "" {
+		return trace.BadParameter("missing file name for %q test", r.Name)
+	}
+	return nil
+}
+
+// Flags returns command-line flags for this fio job spec.
+func (r FioJobSpec) Flags() (flags []string) {
+	if r.Name != "" {
+		flags = append(flags, fmt.Sprint("--name=", r.Name))
+	}
+	if r.ReadWrite != "" {
+		flags = append(flags, fmt.Sprint("--rw=", r.ReadWrite))
+	}
+	if r.IoEngine != "" {
+		flags = append(flags, fmt.Sprint("--ioengine=", r.IoEngine))
+	}
+	if r.Fdatasync {
+		flags = append(flags, "--fdatasync=1")
+	}
+	if r.Filename != "" {
+		flags = append(flags, fmt.Sprint("--filename=", r.Filename))
+	}
+	if r.Size_ != "" {
+		flags = append(flags, fmt.Sprint("--size=", r.Size_))
+	}
+	if r.Runtime != nil {
+		flags = append(flags, fmt.Sprintf("--runtime=%vs", r.Runtime.GetSeconds()))
+	}
+	return flags
+}
+
+// GetWriteIOPS returns number of write iops.
+func (r FioJobResult) GetWriteIOPS() float64 {
+	return r.Write.Iops
+}
+
+// GetFsyncLatency returns 99th percentile of fsync latency in milliseconds.
+func (r FioJobResult) GetFsyncLatency() int64 {
+	return r.Sync.Latency.Percentile[bucket99] / 1000000
```

Comment:
nit: do you need to validate the percentile map before using the `bucket99` slot?

Reply:
Fio always populates this bucket unless you override the percentiles list via a parameter. Probably a better way would be to pass bucket to this method but it should be fine for now.